src/box/memtx_hash.c

```diff
@@ -433,10 +433,24 @@ memtx_hash_index_create_iterator(struct index *base, enum iterator_type type,
 			light_index_iterator_begin(&index->hash_table, &it->iterator);
 			it->base.next = hash_iterator_ge;
 		}
+		/**
+		 * This iterator needs to be supported as a legacy.
+		 * The reason for tracking the gap is the same as below.
+		 */
+		memtx_tx_track_gap(in_txn(), space_by_id(it->base.space_id),
+				   &index->base, NULL, type,
+				   key, part_count);
 		break;
 	case ITER_ALL:
 		light_index_iterator_begin(&index->hash_table, &it->iterator);
 		it->base.next = hash_iterator_ge;
+		/**
+		 * Mark gap from some unknown point to infinity as visited
+		 * in order to go to read view in case of conflict.
+		 */
+		memtx_tx_track_gap(in_txn(), space_by_id(it->base.space_id),
```

Comment:
Here, and above. I seems that it's senseless to pass valid key and part_count here.

Reply:
Thanks a lot, as I can see, it's even crucial here to pass `NULL` as a key as far as the `memtx_tx_handle_gap_write()` handles only this case in a way we need (marking the gap for splitting but not actually splitting it)
```cpp
rlist_foreach_entry_safe(item, list, in_nearby_gaps, tmp) {
	...
	if (item->key == NULL) {
		if (memtx_tx_cause_conflict(txn, item->txn) != 0)
			return -1;
		is_split = true;
		continue;
	} else ...
}
```
Fixed.

Reply:
Much better now, but there's a small problem left. You see, we'll get `is_split = true` in out case. That means that we'll split this 'interval' into two parts - to the left of the inserted key and to the right. But I'm sure that we shouldn't - there can be no intervals in case of hash index.
I think we should mark this gap in hash index somehow - for example with part_count = UINT32_MAX - and don't split in that case.

Reply:
Rebased after #6252 and fixed.